### ISSUE TITLE
Allow require() but forbid module.exports

### ISSUE THREAD

**Issue by Wilfred** [benmosher/eslint-plugin-import#548]
I'm porting a codebase to ES6. In some cases we explicitly want to use `require()`, because `import` must be at the top of the file.

For example, we sometimes write:

``` js
if (process.env.NODE_ENV == "production") {
  require('../foo.css');
}
```

or in tests we sometimes want to only require things at certain points. We also conditionally require polyfills.

Could we add a configuration option to allow `require()` inside `no-commonjs`? Alternatively, would it make sense to split the rule into `no-commonjs-require` and `no-commonjs-exports`?

**Comment by benmosher** [comment]
Are you doing so frequently enough that 

```js
if (process.env.NODE_ENV == "production") {
  require('../foo.css'); // eslint-disable-line import/no-commonjs
}
```

would be burdensome?

**Comment by Wilfred** [comment]
Yep, I think so. Our current workaround is (ab)using `no-restricted-globals` to simply blacklist `module`.

**Comment by benmosher** [comment]
what if `require` were only allowed when not at top-level scope, behind a flag?

i.e. you can set `allowDynamicRequire: true` and it would allow your example, but top-level `require`s would still be reported?

```js
// allowDynamicRequire: true
const fs = require('fs') // still reported
if (process.env.NODE_ENV == "production") {
  require('../foo.css'); // fine
}
```

**Comment by Wilfred** [comment]
That would cover most cases, but we still get tripped up over this:

```js
if (!window.Promise) {
  require('./promise-shim');
}

// Load Foo *after* we have shim in place.
const Foo = require('./Foo').default;
```

**Comment by jfmengels** [comment]
@benmosher I don't mind having this, but this should then probably be included in v2. At least if we want to split the rule into two and then remove the original one.

Either that or have an option that allows either `require`, which might be a bit simpler.

**Comment by benmosher** [comment]
Yeah, I'm leaning towards flags on this. So probably no breaking changes, I still like the default rule just reporting on all CJS usage, regardless of whatever modifications/new rules.

**Comment by jfmengels** [comment]
Agreed on the flags and defaults.
Let's just add a flag to allow `require()` for now, and wait for the `module.exports` until someone requests it with good reason. Would that work for you @Wilfred?

**Comment by Wilfred** [comment]
Yep, that would work for me.

**Comment by benmosher** [comment]
actually I'm thinking the setting should be

```js
require: "allow" | "deny" (default) | "dynamic"
```

@Wilfred you could use `"allow"` to support both your cases, but I also like having a mode that allows dynamic requires (I would probably use that at work actually).

**Comment by benmosher** [comment]
also not super-attached to the word `dynamic`, would be up for bikeshedding a more obvious name for that mode if anyone has ideas.

**Comment by jfmengels** [comment]
@benmosher What would dynamic be? A require with an expression as the argument, or one in a conditional statement/other construct?

**Comment by benmosher** [comment]
> one in a conditional statement/other construct

i.e. 

```js 
const someVariedImport = (someCondition) ? require('one') : require('other');
```

**Comment by benmosher** [comment]
Expression as the argument would be valid too, I suppose. I've just never needed to do that. I'm not sure that this rule would even trigger on that ATM anyway, I think it looks for a single string literal argument.

**Comment by jfmengels** [comment]
dynamic is used as the term for no-dynamic-require, maybe that's not a good name too.

Where I see requires the most in the codebases, is with hot-reloading, so
```js
import foo from './foo';
if (module.hot) {
  const foo = require('./foo');
  //...
}
```

> I'm not sure that this rule would even trigger on that ATM anyway

Yes, that looks like what it is doing.

**Comment by benmosher** [comment]
@[potential implementer]: don't feel like you have to implement `dynamic` mode, a PR implemented `allow` / `deny` would be great too.
